File: crates/agent/src/ask.rs
//! ASKING — everything that decides what ONE model call contains: the phase's
//! configuration, the toolbox that phase grants, the Document assembled under
//! its budget, and the contract demanded back.
//!
//! It is apart from `step.rs`, which owns the TRANSITIONS, because these
//! functions have one obligation to each other that no transition shares: what
//! the model may call and what it is TOLD it may call must be the same set
//! (I13). That is a property of these five together, and it is checkable by
//! reading them together.

use context::ProviderFormat;
use kernel::EndpointName;

use crate::components;
use crate::components::ResponseContract as Contract;
use crate::effect::Effect;
use crate::paper;
use crate::phase::{v1_phases, PhaseConfig, ResponseContract};
use crate::state::AgentState;
use crate::toolbox::Toolbox;

/// This agent's phase configuration.
pub(crate) fn config(state: &AgentState) -> PhaseConfig {
    v1_phases()
        .into_iter()
        .find(|c| c.phase == state.phase)
        .expect("current phase is configured")
}

/// The toolbox this phase grants — the ONLY source of what the model is told
/// it can call: this agent's own toolbox (`subagent::toolbox_for`) narrowed
/// by the phase's `ToolScope`.
pub(crate) fn scoped_tools(state: &AgentState, cfg: &PhaseConfig) -> Toolbox {
    // …AND NARROWED AGAIN BY THE STAGE (20). `plan` and `critique` are told in
    // words to call nothing; this is what makes the words true. Enforcing it
    // here rather than trusting the brief is `engine: base`'s lesson — a
    // capability described but not enforced is a setting that looks applied.
    let stage = crate::stages::current(state);
    // …AND THE PLAN STAGE GETS EXACTLY WHAT ITS BRIEF NAMES. It is told to list
    // the skills and read the ones that apply, so refusing it every tool would
    // make that instruction a lie, and granting it the whole toolbox would let
    // it start the work it is supposed to be planning. Two tools, named here
    // and nowhere else.
    if crate::brief::skill_only(stage) {
        let skills = [crate::skills::LIST_SKILLS, crate::skills::READ_SKILL]
            .iter()
            .map(|n| kernel::ToolId((*n).into()))
            .collect();
        return state.toolbox.scoped(&crate::phase::ToolScope::Only(skills));
    }
    match crate::stages::tools_on(stage) {
        true => state.toolbox.scoped(&cfg.tools),
        false => Toolbox::default(),
    }
}

/// Assemble the paper and ask the model. The affordances and response-contract
/// sections are rewritten from the phase's granted toolbox first, so what the
/// model may call and what it is told it may call cannot drift (I13).
pub(crate) fn call_model(state: &mut AgentState, at: kernel::Timestamp) -> Effect {
    let cfg = config(state);
    let tools = scoped_tools(state, &cfg);
    // WHO IS READING. This is the only place that holds both the endpoint and
    // the paper, so it is the only place that can choose the notation the
    // paper is written in (I13). Set BEFORE the rebuilds below, because each
    // one reads the request off the paper.
    // G4 target: the local OpenAI-compatible proxy, text-only.
    let format = ProviderFormat::OpenAiChat {
        vision: false,
        audio: false,
    };
    state.paper.form = context::Form::for_target(format);
    // The sections rebuilt for every single call — WHICH ones, and why, is
    // `components::dynamic`'s to say. This walks the set; it does not name it.
    for block in components::dynamic(state, &cfg, &tools, at) {
        paper::set_component(&mut state.paper, block.as_ref(), at);
    }
    // …AND THEN THE ONES WHOSE WORDS NAME TOOLS, AGAIN, HONESTLY (T25).
    for block in per_stage_blocks(state, &tools) {
        paper::set_component(&mut state.paper, &block, at);
    }
    Effect::CallModel {
        document: context::assemble(&state.paper, state.phase, cfg.budget),
        format,
        endpoint: EndpointName("model".into()),
        model: state.model.clone(),
        temperature: state.temperature,
        speaker: String::new(), // this agent's own turn
    }
}

/// THE FACULTY BLOCKS WHOSE PROSE NAMES TOOLS, RE-RENDERED AGAINST THE GRANT
/// THIS CALL ACTUALLY HAS (T25).
///
/// THE DEFECT. `## space` names `observe`, `find_files` and `start_process`
/// when the agent holds them — a fix that made the block honest per AGENT. But
/// a host writes that block from `AgentState.toolbox`, and what a TURN may call
/// is [`scoped_tools`], narrowed by the stage. On shipped `main` in `strategy`
/// — the first call of every single turn — `## affordances` said "No tools are
/// installed" and `## space` named three of them five lines later.
///
/// WHY THE SEAM IS HERE AND NOT IN THE HOST. `core::faculty::run::refresh_all`
/// runs ONCE per pass, before an event is pumped, and it is async. The stage
/// cursor moves INSIDE that pump: a stage's reply advances `state.stage` and
/// builds the next call in the same synchronous `step`. So a block written by a
/// host is always at least one stage stale, and handing the host a stage-scoped
/// toolbox would only move the lie rather than remove it. The grant is decided
/// here, synchronously, by the function above; the words that name it belong
/// beside the decision. That is this file's stated obligation — what the model
/// may call and what it is TOLD it may call must be the same set.
///
/// WHAT WAS REJECTED. (1) Scoping `faculty::Sensing.tools` — stale, above.
/// (2) Predicting the next stage in `refresh_all` — a second copy of
/// `stages::next` living in the runtime. (3) Censoring tool names out of
/// rendered parts at assembly — text surgery on prose the core did not write.
/// (4) A `fn(&Toolbox)` hook on `faculty::Faculty` so any faculty could declare
/// grant-dependent prose — generality with one user today; the class test in
/// `tests/prompt.rs` is what makes the SECOND user fail loudly instead.
///
/// It writes to the PAPER and never back into `senses`: `senses` is the record
/// of what a host last saw, read elsewhere as exactly that, and the paper is
/// the per-call artifact. `set_component` upserts by id, so this replaces the
/// host's wording for this call only.
fn per_stage_blocks(state: &AgentState, tools: &Toolbox) -> Vec<components::Sensed> {
    crate::faculty::blocks_of(&state.faculties)
        .into_iter()
        .filter(|block| block.id == crate::faculty::SPACE)
        .map(|block| components::Sensed {
            block,
            parts: components::space_parts(&state.space, tools),
        })
        .collect()
}

/// What the phase demands back, as the component that will render it.
///
/// The envelope form is offered only when there is something to call: telling
/// a model it may call tools and then showing it none is an invitation to
/// invent one.
pub(crate) fn contract(cfg: &PhaseConfig, tools: &Toolbox) -> Contract {
    match (cfg.contract, tools.is_empty()) {
        (ResponseContract::ToolEnvelope, false) => Contract::tool_envelope(),
        _ => Contract::prose(),
    }
}

/// …UNLESS THE STAGE DEMANDS A SHAPE (the strategy vote). A stage whose reply
/// the machine PARSES states the reply as fields; the phase's contract is what
/// every other stage falls back to.
pub(crate) fn stage_contract(state: &AgentState, cfg: &PhaseConfig, tools: &Toolbox) -> Contract {
    crate::brief::contract(crate::stages::current(state))
        .unwrap_or_else(|| contract(cfg, tools))
}

File: packages/agent/src/ask.js
/**
 * ASKING — everything that decides what ONE model call contains: the toolbox
 * this stage grants, the components rebuilt against that grant, the paper
 * assembled under the budget this model's card derives, and the contract
 * demanded back.
 *
 * It is apart from `step.js`, which owns the TRANSITIONS, because these
 * functions have one obligation to each other that no transition shares: WHAT
 * THE MODEL MAY CALL AND WHAT IT IS TOLD IT MAY CALL MUST BE THE SAME SET
 * (I13). One granted toolbox feeds both the affordances block and the response
 * contract, so the two cannot drift — and that is checkable by reading these
 * forty lines together.
 *
 * THE PAPER IS DERIVED, NEVER MUTATED IN PLACE. `paperFor` returns a new source
 * list every call; the state's own paper is read and not written. The
 * predecessor upserted into a long-lived `State`, so a block written for turn N
 * was still in the prompt for turn N+1 — which is how `## space` named three
 * tools the running stage had not been granted.
 *
 * WHERE THE IDENTITY COMES FROM. The soul is written HERE, from `state.prompt`,
 * on every call — it is not seeded once and left standing, because the prompt a
 * person edits mid-run must reach the next call and not the one after it. What
 * this file does not write is the conversation: the exchanges are the paper's
 * own, carried across turns, and `paperFor` reads them without touching them.
 * @module
 */

import { MODEL_ENDPOINT } from '@harness/kernel'
import { adapterFor, assemble, budgetFor, sectionOf } from '@harness/context'
import { callModel } from './effect.js'
import { endTurn } from './ending.js'
import { facultyBlocks } from './faculty/index.js'
import { affordances, contract, directive, observations, sensed, soul, taskBlock } from './paper.js'
import { backoffMs } from './retry.js'
import { grant, resolveStage } from './stages.js'
import { usages } from './toolbox.js'

/** @typedef {import('@harness/context').ModelCard} ModelCard */
/** @typedef {import('@harness/context').SectionSource} SectionSource */
/** @typedef {import('@harness/context').Component} Component */
/** @typedef {import('@harness/kernel').Timestamp} Timestamp */
/** @typedef {import('./effect.js').Effect} Effect */
/** @typedef {import('./stages.js').Stage} Stage */
/** @typedef {import('./state.js').AgentState} AgentState */
/** @typedef {import('./tools.js').Tool} Tool */

/** What one call is assembled against: the stage it is taken in, the card that derives its budget, and the backoff a retry carries. @typedef {{stage: Stage, card: ModelCard, at: Timestamp, afterMs?: number}} Asking */

/**
 * THE TOOLBOX THIS CALL ACTUALLY HAS — this agent's own, narrowed by the
 * stage's allowlist, and the only source of what the model is told it can call.
 * A stage scoped to `none` yields an empty array, which is why a stage that may
 * not act cannot even NAME a tool: the affordances block is built from this.
 * @param {AgentState} state @param {Stage} stage @returns {Tool[]}
 */
export function granted(state, stage) {
  return grant(stage.toolAllowlist, state.toolbox)
}

/**
 * THE PAPER FOR THIS TURN, derived. Whatever the paper already carries, with
 * the components this call owns written over it by id.
 *
 * Upsert and not append: a block the paper has never held is added, which is
 * what opens the prompt to a faculty that was declared in a file rather than
 * compiled in. Ordering is structural — `assemble` sorts by slot and nothing
 * else — so a source's position here never reaches the model.
 * @param {AgentState} state @param {Asking} of @returns {SectionSource[]}
 */
export function paperFor(state, of) {
  const held = /** @type {SectionSource[]} */ ([...state.paper.sources])
  const rebuilt = components(state, of)
  for (const component of rebuilt) {
    const source = { section: sectionOf(component, of.at), summary: null }
    const at = held.findIndex((s) => s.section.id === source.section.id)
    if (at === -1) held.push(source)
    else held[at] = source
  }
  return held
}

/**
 * THE BLOCKS THIS CALL OWNS, in prompt order. Order here is documentation and
 * not mechanism, because assembly sorts by slot; they are listed in the order
 * the model reads them anyway, so nobody has to consult the slot table to
 * picture the result.
 *
 * A block belonging to a CAPABILITY is not listed here and adding one does not
 * mean editing this function: the agent file names a faculty, the faculty
 * declares its block, and a host's most recent parts render into it.
 * @param {AgentState} state @param {Asking} of @returns {Component[]}
 */
function components(state, of) {
  const tools = granted(state, of.stage)
  return [
    soul(state.prompt),
    affordances(usages(tools)),
    ...facultyBlocks(state.faculties).map((block) => sensed(block, state.senses[block.id] ?? [])),
    taskBlock(state.task),
    observations(state.observations),
    directive(of.stage.brief),
    contract({ hasTools: tools.length > 0, schema: of.stage.responseSchema }),
  ]
}

/**
 * ASSEMBLE THE PAPER AND ASK THE MODEL.
 *
 * The budget is DERIVED from the card and never declared: the window belongs to
 * the model and what the paper may spend is what is left after the reply and
 * the estimator's slack come out of it. One constant for every model is how
 * 8192 survived a measurement that said 4174 tokens did not fit in 4096.
 *
 * THE IMAGE ARITHMETIC IS THE CARD'S PROVIDER, NOT OPENAI'S. Anthropic bills a
 * photograph at about w*h/750 and the tile rule understates that by ~3x, so a
 * paper fitted under the wrong rule fits a window it will not fit — and the
 * receipt then names a provider that was never asked. `adapterFor(card.kind)`
 * is the same table `assemble` names in its own doc comment.
 * @param {AgentState} state @param {Asking} of @returns {Effect}
 * @throws {HarnessError} by law name — `elided_but_named`, `window_too_small`, `unknown_provider`
 */
export function askModel(state, of) {
  const document = assemble(
    // `answer` is a stage this loop walks (`STAGES_OF.answer`) and the kernel's
    // `StageId` does not name it. Filed as a cross-lane request; the cast is
    // here rather than in the vocabulary because widening a shared dictionary
    // is not this lane's to do.
    { stage: /** @type {import('@harness/kernel').StageId} */ (of.stage.name), sources: paperFor(state, of) },
    budgetFor(of.card),
    adapterFor(of.card.kind).images,
  )
  return callModel({
    turnId: state.turnId,
    document,
    endpoint: MODEL_ENDPOINT,
    model: state.model,
    temperature: state.temperature,
    afterMs: of.afterMs ?? 0,
  })
}

/**
 * WHICH STAGE THIS CALL IS TAKEN IN. A turn walking no stage list is the bare
 * react loop — every agent written before the key existed — and it takes
 * `work`, which is briefed by the person's own message and carries no brief of
 * its own. The cursor past the end of the list is the same case.
 *
 * A BRIEFED STAGE WHOSE FILE NEVER LOADED REFUSES. A stage entered with no
 * instruction writes nothing while looking exactly like one that ran.
 * @param {AgentState} state @returns {{stage: Stage} | {problem: string}}
 */
export function stageNow(state) {
  const name = state.stages[state.stage] ?? 'work'
  const resolved = resolveStage(/** @type {import('./strategy.js').StageName} */ (name), {
    briefs: state.briefs,
    hasSpace: state.space !== null,
  })
  return 'refusal' in resolved ? { problem: resolved.refusal.message } : resolved
}

/**
 * THE NEXT MODEL CALL, OR THE SENTENCE SAYING WHY THERE ISN'T ONE.
 *
 * ASSEMBLY THROWS BY LAW NAME AND THIS TURNS IT INTO AN ENDING. A section elided
 * while another names it, a window too small for its own reply, a card naming a
 * provider nobody implements — each is a real, nameable state of THIS agent,
 * and a turn that cannot be assembled must end saying which law refused it. Letting it throw past the
 * reducer would take the whole page down over one agent's paper.
 * @param {AgentState} state @param {Timestamp} at @param {number} [afterMs]  the backoff a retry carries
 * @returns {{effect: Effect} | {problem: string}}
 */
export function askFor(state, at, afterMs = 0) {
  if (!state.card) return { problem: `no catalogue entry named "${state.model}", so there is no window to assemble against` }
  const stage = stageNow(state)
  if ('problem' in stage) return stage
  try {
    return { effect: askModel(state, { stage: stage.stage, card: state.card, at, afterMs }) }
  } catch (cause) {
    return { problem: cause instanceof Error ? cause.message : 'the paper could not be assembled' }
  }
}

/**
 * ASK THE MODEL — the ONE call site, and the transition it hands back.
 *
 * A CALL THAT CANNOT BE ASSEMBLED ENDS THE TURN, saying which law refused it.
 * Throwing past the reducer would take the page down over one agent's paper,
 * and swallowing it would leave the turn awaiting a call nobody described.
 * @param {AgentState} state @param {Timestamp} at @param {number} [attempts]  how many of this turn's calls have already failed
 * @returns {{state: AgentState, effects: Effect[]}}
 */
export function nextCall(state, at, attempts = 0) {
  /** @type {AgentState} */
  const asking = { ...state, attempts }
  const asked = askFor(asking, at, attempts === 0 ? 0 : backoffMs(attempts))
  if ('problem' in asked) return endTurn(asking, asked.problem)
  return { state: asking, effects: [asked.effect] }
}
